fix(io): allow unsafe addresses

dtfiedler opened the pull request:
We allow optionally providing `Allow-Unsafe-Addresses` flag to `transfer` and `vaultedTransfer`. This gives flexibility to future signature types but provides some contract enforcement of address types by default. Enforcement of other address-like things - like `processId`'s - are relaxed. 

This also removes the protection of names that look like wallet addresses, but I can be convinced to leave that in.

Commit:
fix(io): open up allowed address types

Commit:
fix(io): add support for optional `Allow-Unsafe-Addresses` tag

This tag allows overriding the default behavior of unsafe transferring.

dtfiedler commented on src/arns.lua:
```diff
@@ -474,16 +474,21 @@ function arns.calculateYearsBetweenTimestamps(startTimestamp, endTimestamp)
-	assert(type(name) == "string", "Name is required and must be a string.")
-	assert(#name >= 1 and #name <= 51, "Name pattern is invalid.")
-	assert(name:match("^%w") and name:match("%w$") and name:match("^[%w-]+$"), "Name pattern is invalid.")
-	assert(not utils.isValidAOAddress(name), "Name cannot be a wallet address.")
```
we added this in, but bc we want to support more sig types in the future, it seems this should just be removed. fine to add back if you think otherwise @arielmelendez / @vilenarios

dtfiedler commented on src/main.lua:
```diff
@@ -1206,8 +1215,8 @@ end)
 	local vaultId = msg.Tags["Vault-Id"]
-	assert(utils.isValidAOAddress(gatewayAddress), "Invalid gateway address")
-	assert(utils.isValidAOAddress(vaultId), "Invalid vault id")
+	assert(utils.isValidAddress(gatewayAddress), "Invalid gateway address")
```
these are laxed bc to join the network `msg.From` will be used for the gateway address. if AO introduces support for new addresses, we don't want to worry about blocking delegations on those new addresses. the check for the `gateway` itself is sufficient.

arielmelendez replied:
Is it actually relaxed if you're not passing "allow unsafe" here?

dtfiedler replied:
🤔

dtfiedler commented on src/main.lua:
```diff
@@ -1427,10 +1436,10 @@ addEventingHandler(ActionMap.ReassignName, utils.hasMatchingTag("Action", Action
 	local timestamp = msg.Timestamp
 	assert(name and #name > 0, "Name is required")
-	assert(utils.isValidAOAddress(newProcessId), "Process Id must be a valid AO signer address..")
+	assert(utils.isValidAddress(newProcessId), "Process Id must be a valid address.")
```
similarly - process ids can be flexible - maybe integration with EVM smart contracts will be available in the future 🤷

arielmelendez replied:
Is it flexible if allow unsafe is not passed here?

dtfiedler replied:
yeah you're right. frick me.

dtfiedler replied:
i updated the call sig to not allow optionally, and set `true` where want to allow it to be relaxed.

[5601bc4](https://github.com/ar-io/ar-io-network-process/pull/216/commits/5601bc436f37046b22fd2ff71e366efc599fcb30)

lmk if you reason about it differently - my brain turned into spaghetti as we discussed the correct defaults for this

arielmelendez commented on src/main.lua:
```diff
@@ -540,6 +548,7 @@ addEventingHandler(ActionMap.Transfer, utils.hasMatchingTag("Action", ActionMap.
 			Action = "Debit-Notice",
 			Recipient = recipient,
 			Quantity = msg.Tags.Quantity,
+			["Allow-Unsafe-Addresses"] = tostring(allowUnsafeAddresses),
```
Is this logically significant?

dtfiedler replied:
[e7e4e0a](https://github.com/ar-io/ar-io-network-process/pull/216/commits/e7e4e0a6e5d0305401b706cd5d3b8be13764628e)

dtfiedler committed:
chore: add explanation of arns regex

Co-authored-by: Derek Sonnenberg <derek@ardrive.io>

fedellen reviewed (dismissed):
I like the compromise here. Other iteration seemed a little dangerous for funds being lost due to typos or incomplete commands, although client-side could've still enforced address validation.

Commit:
chore(tags): remove tostring for booleans

These get encoded by default, so redundant

Commit:
fix(io): make sure to provide the tag everywhere, and R.E.L.A.X. the constraint where sensible

dtfiedler committed:
Merge branch 'develop' into PE-7160-allow-all-addresses

Commit:
fix(delegates): remove vaults object from get delegates handler

arielmelendez commented on src/arns.lua:
```diff
@@ -474,16 +474,21 @@ function arns.calculateYearsBetweenTimestamps(startTimestamp, endTimestamp)
+--- @param name string The name to check
+function arns.assertValidArNSName(name)
+	assert(type(name) == "string", "Name is required and must be a string.")
+	assert(#name >= constants.MIN_NAME_LENGTH and #name <= constants.MAX_NAME_LENGTH, "Name pattern is invalid.")
```
Nit: could have a more descriptive error. Same for the regex pattern below. Could even have the regex string in the error message.

dtfiedler replied:
[f4db59a](https://github.com/ar-io/ar-io-network-process/pull/216/commits/f4db59a3098c6ab86484ec7beeacb9d08412cd92)

arielmelendez commented on src/arns.lua:
```diff
@@ -499,7 +504,7 @@ function arns.assertValidBuyRecord(name, years, purchaseType, processId)
 	-- assert processId is valid pattern
 	assert(type(processId) == "string", "Process id is required and must be a string.")
-	assert(utils.isValidAOAddress(processId), "Process Id must be a valid AO signer address..")
+	assert(utils.isValidAddress(processId, true), "Process Id must be a valid address.")
```
Nit: could make a `utils.isValidArNSTarget` function to consolidate the logic here.

arielmelendez commented on src/constants.lua:
```diff
@@ -14,6 +14,14 @@ constants.minimumWithdrawalAmount = constants.mIOPerIO -- the minimum amount tha
+-- - Can contain alphanumeric characters and hyphens (%w-)
+-- - Ends with an alphanumeric character (%w)
+-- - Does not allow names to start or end with a hyphen
+constants.ARNS_NAME_REGEX = "^%w[%w-]*%w$"
```
This regex does not allow 1 character names!

arielmelendez replied:
I think it could be:
```
^%w[%w-]*%w?$
```

arielmelendez replied:
Is there not a unit test ensuring that 1 character names are allowed?

dtfiedler replied:
ha, looks like that is in fact missing. will add and reasses

Commit:
fix(arns): modify regex to allow one char name

dtfiedler committed:
Merge branch 'develop' into PE-7160-allow-all-addresses

Commit:
feat(redelegation): prune redelgations opportunistically PE-7147

Commit:
test(redelegation): test prune skipping PE-7147

Commit:
fix(gar): do not distribute rewards to delegate that has 0 stake

Commit:
feat(epochs): using next based iteration during pruning PE-7147

Commit:
fix(gateways): remove delegates and vaults from gateways

This will minimize the gateway request. The SDK and downstream apps are already using the newest APIs

Commit:
chore(test): update monitor test to use latest sdk

Commit:
fix(primary): add handler for fetching a primary name request, fix request primary name tag

fedellen committed:
fix: assert required years on lease token cost

Commit:
feat(epochs): prune epochs opportunistically PE-7147

Commit:
feat(epochs): keep old setting name PE-7147

Commit:
fix(gar): update gar settings, minimum stakes, withdrawal lengths

Moves the gar functions to helpers.mjs, and updates tests with new values

Includes PE-7186/PE-7187/PE-7188

Commit:
Merge remote-tracking branch 'origin/develop' into PE-7160-allow-all-addresses

Commit:
chore(deps): fix deps

Commit:
Merge remote-tracking branch 'origin/develop' into PE-7160-allow-all-addresses

dtfiedler committed:
Merge branch 'develop' into PE-7160-allow-all-addresses

arielmelendez commented on src/arns.lua:
```diff
@@ -512,7 +524,7 @@ function arns.assertValidBuyRecord(name, years, purchaseType, processId)
 	-- assert processId is valid pattern
 	assert(type(processId) == "string", "Process id is required and must be a string.")
-	assert(utils.isValidAOAddress(processId), "Process Id must be a valid AO signer address..")
+	assert(utils.isValidAddress(processId, true), "Process Id must be a valid address.")
```
I think this might be one where we want to force users to specify allowing unsafe because if they get this wrong, they can brick the name. e.g. they wouldn't be able to use the ANT to reassign to another process ID.

arielmelendez replied:
f602db8

arielmelendez commented on src/arns.lua:
```diff
@@ -1058,7 +1070,7 @@ end
 	assert(record, "Name is not registered")
 	assert(currentTimestamp, "Timestamp is required")
-	assert(utils.isValidAOAddress(newProcessId), "Invalid Process-Id")
+	assert(utils.isValidAddress(newProcessId, true), "Invalid Process-Id") -- allow unsafe addresses for processId
```
Same comment here - we should force user to specify unsafe.

arielmelendez replied:
e1f740d

arielmelendez commented on src/main.lua:
```diff
@@ -572,15 +581,17 @@ addEventingHandler(ActionMap.Transfer, utils.hasMatchingTag("Action", ActionMap.
 			Recipient = recipient,
 			Quantity = tostring(quantity),
-			Data = "You transferred " .. tostring(quantity) .. " to " .. recipient,
+			["Allow-Unsafe-Addresses"] = allowUnsafeAddresses,
```
Does this need to be converted back to a string first?

arielmelendez replied:
616e10d

arielmelendez commented on src/main.lua:
```diff
@@ -572,15 +581,17 @@ addEventingHandler(ActionMap.Transfer, utils.hasMatchingTag("Action", ActionMap.
 			Sender = msg.From,
 			Quantity = tostring(quantity),
-			Data = "You received " .. tostring(quantity) .. " from " .. msg.From,
+			["Allow-Unsafe-Addresses"] = allowUnsafeAddresses,
```
Same comment here about converting back to a string first.

arielmelendez replied:
616e10d

arielmelendez commented on src/main.lua:
```diff
@@ -678,7 +690,7 @@ addEventingHandler(ActionMap.VaultedTransfer, utils.hasMatchingTag("Action", Act
 		Recipient = recipient,
 		Quantity = quantity,
-		Tags = { Action = "Debit-Notice", ["Vault-Id"] = msgId },
+		Tags = { Action = "Debit-Notice", ["Vault-Id"] = msgId, ["Allow-Unsafe-Addresses"] = allowUnsafeAddresses },
```
Same issue here with converting back to string.

arielmelendez replied:
616e10d

arielmelendez commented on src/main.lua:
```diff
@@ -756,15 +769,13 @@ addEventingHandler(ActionMap.BuyRecord, utils.hasMatchingTag("Action", ActionMap
-	assert(type(processId) == "string", "Process id is required and must be a string.")
-	assert(utils.isValidAOAddress(processId), "Process Id must be a valid AO signer address..")
+	arns.assertValidArNSName(name)
+	assert(utils.isValidAddress(processId, true), "Process Id must be a valid address.")
```
Mentioned elsewhere that this should for user to opt into unsafe since this could brick a name.

arielmelendez replied:
e1f740d

arielmelendez commented on src/main.lua:
```diff
@@ -1460,10 +1470,10 @@ addEventingHandler(ActionMap.ReassignName, utils.hasMatchingTag("Action", Action
 	local timestamp = msg.Timestamp
 	assert(name and #name > 0, "Name is required")
-	assert(utils.isValidAOAddress(newProcessId), "Process Id must be a valid AO signer address..")
+	assert(utils.isValidAddress(newProcessId, true), "Process Id must be a valid address.")
```
Mentioned elsewhere, this should be OPT IN to unsafe to avoid bricking ANT.

arielmelendez commented on src/main.lua:
```diff
@@ -2019,9 +2029,9 @@ addEventingHandler("releaseName", utils.hasMatchingTag("Action", ActionMap.Relea
-	assert(name and #name > 0, "Name is required")
-	assert(processId and utils.isValidAOAddress(processId), "Process-Id is required")
-	assert(initiator and utils.isValidAOAddress(initiator), "Initiator is required")
+	assert(name and #name > 0, "Name is required") --- this could be an undername, so we don't want to assertValidArNSName
```
Nit: don't need three hyphens

arielmelendez replied:
b2da18c

arielmelendez commented on src/main.lua:
```diff
@@ -2019,9 +2029,9 @@ addEventingHandler("releaseName", utils.hasMatchingTag("Action", ActionMap.Relea
+	assert(name and #name > 0, "Name is required") --- this could be an undername, so we don't want to assertValidArNSName
+	assert(processId and utils.isValidAddress(processId, true), "Process-Id must be a valid address")
+	assert(initiator and utils.isValidAddress(initiator, true), "Initiator is required")
 	assert(record, "Record not found")
```
Perhaps we shouldn't be fetching the record before validating the name.

arielmelendez replied:
b2da18c

arielmelendez commented on src/main.lua:
```diff
@@ -2197,8 +2207,8 @@ addEventingHandler("auctionBid", utils.hasMatchingTag("Action", ActionMap.Auctio
-	assert(bidder and utils.isValidAOAddress(bidder), "Bidder is required")
-	assert(processId and utils.isValidAOAddress(processId), "Process-Id is required")
+	assert(bidder and utils.isValidAddress(bidder, true), "Bidder is required")
+	assert(processId and utils.isValidAddress(processId, true), "Process-Id must be a valid Arweave address")
```
Should be opt in to unsafe to avoid bricking names.

arielmelendez replied:
b2da18c

arielmelendez commented on src/utils.lua:
```diff
@@ -246,7 +246,7 @@ end
 --- Checks if an address is a valid Arweave address
 --- @param address string The address to check
---- @return boolean isValidArweaveAddress - whether the address is a valid Arweave address
+--- @return boolean isValidAddress - whether the address is a valid Arweave address
```
No need for a variable name in the return doc. Just use a hashtag.

arielmelendez replied:
b2da18c

arielmelendez commented on src/utils.lua:
```diff
@@ -258,11 +258,27 @@ function utils.isValidEthAddress(address)
+		return false
+	end
+	local match = string.match(address, "^[%w_-]+$")
+	return match ~= nil and #address >= 1 and #address <= 128
```
Nit: use constants

arielmelendez replied:
b2da18c

arielmelendez commented on src/utils.lua:
```diff
@@ -258,11 +258,27 @@ function utils.isValidEthAddress(address)
-function utils.isValidAOAddress(url)
-	return url and (utils.isValidArweaveAddress(url) or utils.isValidEthAddress(url)) or false
+--- @param address string|nil The address to check
+--- @param allowUnsafe boolean Whether to allow unsafe addresses, defaults to false
```
type is `boolean|nil`

arielmelendez replied:
c3992b1

arielmelendez commented on src/vaults.lua:
```diff
@@ -52,8 +51,10 @@ end
 --- @param lockLengthMs number The lock length in milliseconds
 --- @param currentTimestamp number The current timestamp
 --- @param vaultId string The vault id
+--- @param allowUnsafeAddresses boolean Whether to allow unsafe addresses, since this results in funds eventually being sent to an invalid address
```
Can relax the type to `boolean|nil`

arielmelendez reviewed (changes requested):
Needs a little work. Almost there though.

Commit:
feat(buy record): some enforcement of opt in to purchasing a name with unsafe process id PE-7160

Commit:
feat(buy record): more enforcement of opt in to purchasing or reassigning a name with unsafe process id PE-7160

Commit:
feat(auctions): enforce opt in to unsafe process ids for auction bids PE-7160

Commit:
docs(utils): minor type update PE-7160

Commit:
docs(utils): another minor type update PE-7160

Commit:
feat(unsafe addresses): send back setting as string PE-7160

arielmelendez commented on src/main.lua:
```diff
@@ -689,6 +701,7 @@ addEventingHandler(ActionMap.VaultedTransfer, utils.hasMatchingTag("Action", Act
 		Tags = {
 			Action = ActionMap.CreateVault .. "-Notice",
 			["Vault-Id"] = msgId,
+			["Allow-Unsafe-Addresses"] = allowUnsafeAddresses,
```
616e10d

fedellen commented on src/arns.lua:
```diff
@@ -615,7 +631,7 @@ function arns.getTokenCost(intendedAction)
 		-- stub the process id as it is not required for this intent
 		local processId = "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"
-		arns.assertValidBuyRecord(name, years, purchaseType, processId)
+		arns.assertValidBuyRecord(name, years, purchaseType, processId, false)
```
this blocks getting a token cost for an unsafe name, we could use the allow-unsafe tag in this handler

fedellen replied:
or might be better to have `true` in this read level API

fedellen commented on src/main.lua:
```diff
@@ -756,23 +773,32 @@ addEventingHandler(ActionMap.BuyRecord, utils.hasMatchingTag("Action", ActionMap
-	assert(type(name) == "string" and #name > 0 and #name <= 51 and not utils.isValidAOAddress(name), "Invalid name")
-	assert(type(processId) == "string", "Process id is required and must be a string.")
-	assert(utils.isValidAOAddress(processId), "Process Id must be a valid AO signer address..")
+	arns.assertValidArNSName(name)
+	assert(utils.isValidAddress(processId, true), "Process Id must be a valid address.")
```
```suggestion
	arns.assertValidArNSName(name)
	assert(utils.isValidAddress(processId, allowUnsafeProcessId), "Process Id must be a valid address.")
```

fedellen replied:
I guess not really, as long as it gets asserted again with the boolean passed down in `arns.lua`
